icicle/include/icicle/fields/field.h

```diff
@@ -254,422 +271,23 @@ class Field
-      uint32_t carry2 = add_sub_u32_device<(TLC >> 1), true, true>(b, &b[TLC >> 1], &diffs[TLC >> 1]);
-      // Compute the "middle part" of Karatsuba: \f$ a_{lo} \cdot b_{hi} + b_{lo} \cdot a_{hi} \f$.
-      // This is where the assumption about unset high bit of `a` and `b` is relevant.
-      multiply_and_add_short_raw_device(diffs, &diffs[TLC >> 1], middle_part, r, &r[TLC]);
-      // Corrections that need to be performed when differences are negative.
-      // Again, carry doesn't need to be propagated due to unset high bits of `a` and `b`.
-      if (carry1)
-        add_sub_u32_device<(TLC >> 1), true, false>(&middle_part[TLC >> 1], &diffs[TLC >> 1], &middle_part[TLC >> 1]);
-      if (carry2) add_sub_u32_device<(TLC >> 1), true, false>(&middle_part[TLC >> 1], diffs, &middle_part[TLC >> 1]);
-      // Now that middle part is fully correct, it can be added to the result.
-      add_sub_u32_device<TLC, false, true>(&r[TLC >> 1], middle_part, &r[TLC >> 1]);
-
-      // Carry from adding middle part has to be propagated to the highest limb.
-      for (size_t i = TLC + (TLC >> 1); i < 2 * TLC; i++)
-        r[i] = ptx::addc_cc(r[i], 0);
-    } else if (TLC == 2) {
-      __align__(8) uint32_t odd[2];
-      r[0] = ptx::mul_lo(a[0], b[0]);
-      r[1] = ptx::mul_hi(a[0], b[0]);
-      r[2] = ptx::mul_lo(a[1], b[1]);
-      r[3] = ptx::mul_hi(a[1], b[1]);
-      odd[0] = ptx::mul_lo(a[0], b[1]);
-      odd[1] = ptx::mul_hi(a[0], b[1]);
-      odd[0] = ptx::mad_lo(a[1], b[0], odd[0]);
-      odd[1] = ptx::mad_hi(a[1], b[0], odd[1]);
-      r[1] = ptx::add_cc(r[1], odd[0]);
-      r[2] = ptx::addc_cc(r[2], odd[1]);
-      r[3] = ptx::addc(r[3], 0);
-    } else if (TLC == 1) {
-      r[0] = ptx::mul_lo(a[0], b[0]);
-      r[1] = ptx::mul_hi(a[0], b[0]);
-    }
+    return base_math::template add_sub_limbs<NLIMBS, true, CARRY_OUT>(xs, ys, rs);
   }
 
-#endif // __CUDACC__
   static HOST_DEVICE_INLINE void multiply_raw(const ff_storage& as, const ff_storage& bs, ff_wide_storage& rs)
   {
-#ifdef __CUDA_ARCH__
-    return multiply_raw_device(as, bs, rs);
-#else
-    return host_math::template multiply_raw<TLC>(as, bs, rs);
-#endif
-  }
-
-  static HOST_DEVICE_INLINE void
-  multiply_and_add_lsb_neg_modulus_raw(const ff_storage& as, ff_storage& cs, ff_storage& rs)
-  {
-#ifdef __CUDA_ARCH__
-    return multiply_and_add_lsb_neg_modulus_raw_device(as, cs, rs);
-#else
-    Wide r_wide = {};
-    host_math::template multiply_raw<TLC>(as, get_neg_modulus(), r_wide.limbs_storage);
-    Field r = Wide::get_lower(r_wide);
-    add_limbs<TLC, false>(cs, r.limbs_storage, rs);
-#endif
-  }
-
-  static HOST_DEVICE_INLINE void multiply_msb_raw(const ff_storage& as, const ff_storage& bs, ff_wide_storage& rs)
-  {
-#ifdef __CUDA_ARCH__
-    return multiply_msb_raw_device(as, bs, rs);
-#else
-    return host_math::template multiply_raw<TLC>(as, bs, rs);
-#endif
+    return base_math::template multiply_raw<TLC>(as, bs, rs);
```

Comment:
did you make all these functions inline?

Reply:
yes